harness/determined/keras/_enqueuer.py

```diff
@@ -0,0 +1,355 @@
+    threading.Thread or a multiprocessing.Process; the caller (a _WorkerEnqueuer) is responsible
+    for passing in the appropriate type of Queue.
+
+    Parameters:
+        sequence: the user-provided Keras Sequence
+        queries: a queue of tuples of (indices, order) that need to be read from the sequence
+        answers: a queue of tuples of (data, order) that workers fill with data from the sequence
+    """
+
+    try:
+        while True:
+            query = queries.get()
+            if query is None:
+                return
+            i, order = query
+            data = sequence[i]
+            answers.put((data, order))
+    finally:
+        answers.put(None)
+
+
+class _WorkerEnqueuer(_Enqueuer):
+    """
+    _WorkerEnqueuer defines the semantics for either a threading-based or multiprocessing-based
+    enqueuer (threading and multiprocessing are intentionally designed to be API compatible).  The
+    implementation-specific details are defined in the child classes.
+
+    Generally, the strategy is:
+      - in start(): start a bunch of workers with access to a pair of queues.
+
+      - in data(): fill the query queue with tuples of (index, order) tuples.  The index comes from
+        the _Sampler, and the order just defines in what order we put things into the queue.
+
+        Workers will pop (index, order) tuples fram the query queue and put tuples of (data, order)
+        in the answers queue.
+
+        The main thread will gather the (data, order) tuples, and reorder the data elements to
+        match the order they were requested.  Whenever the next-requested data is available, it
+        will pass yield that data through the generator.
+
+      - in close(): shut down all the workers and clean up resources.
+    """
+
+    def __init__(
+        self,
+        sequence: tf.keras.utils.Sequence,
+        sampler: _Sampler,
+        repeat: bool,
+        workers: int,
+        max_queue_size: int,
+    ):
+        self.sequence = sequence
+        self.sampler = sampler
+        self.repeat = repeat
+        self.max_queue_size = max_queue_size
+        check.gt(max_queue_size, 0, "max_queue_size must be greater than zero")
+
+        # Coordination logic.
+        self.order = 0
+        self.requested = collections.deque()  # type: Deque[int]
+        self.received = {}  # type: Dict[int, Any]
+        self.started = False
+        self.stopped = False
+        self.index_iter = None  # type: Optional[Iterator]
+
+        # Interthread/interprocess communications.
+        self.queries = self.queue_class()()
+        self.answers = self.queue_class()()
+
+        self.workers = [
+            self.worker_class()(target=_worker, args=(self.sequence, self.queries, self.answers))
+            for _ in range(workers)
+        ]
+
+    def start(self) -> None:
+        assert not self.started and not self.stopped, "restarting an enqueuer is not allowed"
```

Comment:
non-blocking: this should probably be a `check`

Reply:
I disagree, as it is an internal object and the assertion is validating code structure, not user inputs, and the check module would not print anything useful anyway.